SQL server AD identity and data encryption support

jaredmoo opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
- Updated `Servers` api version from 2014-04-01 to 2015-05-01-preview, which is SDK compatible and includes support for server managed identity
- Added support for server keys and encryption protectors
- Added support for check server name availability
- Added support for virtual network firewall rules
- Updated server azure ad admin from swagger
- Minor nonfunctional updates to database blob auditing
- **Breaking changes** DatabaseMetrics and ServerMetrics renamed to DatabaeUsage and ServerUsage. These were misleadingly named because metrics is a different API.
- Added database metrics and elastic pool metrics

Note: the server azure ad admin change looks like a breaking change, but the previous version was not yet shipped so it's not breaking.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

jaredmoo committed:
Generate CheckNameAvailability and add tests

jaredmoo committed:
Rerecorded tests for server api-version 2015-05-01.

Also several minor refactorings, removing redundant namespaces, etc.

jaredmoo committed:
Update generate.cmd and bump version to 1.5.0

nathannfan committed:
Add database and elastic pool metrics tests and recordings, re-generate client from latest swagger

mihymel committed:
Adding swagger generated clients for EncryptionProtector and ServerKey operations. Updating the SDK to use a new Server API version (2015-05-01-preview).

jaredmoo committed:
Merge remote-tracking branch 'mihymel/psSdkJson6' into servers_checkNameAvailability

jaredmoo commented:
@mihymel FYI

mihymel commented:
@jaredmoo Thanks for the FYI. Looks good to me.

shahabhijeet commented:
@jaredmoo can you add swagger spec PR link to this PR

shahabhijeet commented on src/SDKs/SqlManagement/Sql.Tests/Sql.Tests.csproj:
```diff
@@ -9,6 +9,13 @@
     <TargetFrameworks>netcoreapp1.1</TargetFrameworks>
   </PropertyGroup>
   <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.KeyVault" Version="2.0.6" />
```
@jaredmoo I was under the impression Nathan Fan updated all SQL Tests to use latest Resource Manager nuget.
I see you are using a really old resource manager to record your tests, please update to the latest (>=1.6.0)

jaredmoo replied:
Yes, Nathan did that. Where are you seeing the old resource manager?

jaredmoo replied:
It look like all the json recordings have `Microsoft.Azure.Management.ResourceManager.ResourceManagementClient/1.6.0.0`

shahabhijeet replied:
@jaredmoo So by default every test project get's 1.1.0 reference of Resource manager.
So unless you explicitly add 1.6.0 to the test project you are still using 1.1.0 resource manager nuget
Here is the link for the change
https://github.com/Azure/azure-sdk-for-net/pull/3272/files#diff-c8ac009921fcb73848451401f0d039f2R16

jaredmoo commented:
There are a few swagger PR combined into this SDK PR:
- https://github.com/Azure/azure-rest-api-specs/pull/1273
- https://github.com/Azure/azure-rest-api-specs/pull/1260
- https://github.com/Azure/azure-rest-api-specs/pull/1251
- https://github.com/Azure/azure-rest-api-specs/pull/1202
- https://github.com/Azure/azure-rest-api-specs/pull/1095

shahabhijeet commented on src/SDKs/SqlManagement/Sql.Tests/AzureRmSqlServerActiveDirectoryAdministratorTest.cs:
```diff
@@ -22,30 +17,14 @@ public void TestSetServerActiveDirectoryAdministrator()
                 // Add new Active Directory Admin
                 ServerAzureADAdministrator newAdmin = new ServerAzureADAdministrator("DSEngAll", new Guid("5e90ef3b-9b42-4777-819b-25c36961ea4d"), new Guid("72f988bf-86f1-41af-91ab-2d7cd011db47"), "sqlcrudtest-3421", "2c647056-bab2-4175-b172-493ff049eb29");
-                ServerAzureADAdministrator createResult = sqlClient.ServerAzureADAdministrators.CreateOrUpdateAzureADAdministrator(resourceGroup.Name, server.Name, newAdmin);
+                ServerAzureADAdministrator createResult = sqlClient.ServerAzureADAdministrators.CreateOrUpdate(resourceGroup.Name, server.Name, newAdmin);
```
@jaredmoo why not use TestFramework to get all your custom key-value pairs injected in your tests.
https://github.com/Azure/azure-powershell/blob/preview/documentation/Using-Azure-TestFramework.md#environment--custom

jaredmoo replied:
Good point, since we didn't get this feedback on the original PR (https://github.com/Azure/azure-sdk-for-net/pull/3196) can we fix this as a separate issue?

jaredmoo committed:
Updated to latest swagger. Adds DB and elastic pool metrics.

jaredmoo committed:
Temporarily removed vnet firewall rules due to test recording difficulties

jaredmoo commented:
Hi @shahabhijeet , thanks for the review so far. I don't think your comments require immediate action. I have updated again to latest swagger - the metrics changes just got merged earlier this afternoon ( https://github.com/Azure/azure-rest-api-specs/pull/1284 ). I also reverted adding the new VNetFirewallRule SDK support for now; we have written tests but need to re-record them and are having temporary difficulties.

jaredmoo committed:
Moved Microsoft.Azure.Management.ResourceManager PackageReference
